### PR TITLE
[#188] Change Categorical to take GenericTensor as value

### PR DESCRIPTION
**Pull request by icantsayrural**

### CODE REVIEW

File: keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Categorical.java [hunk 1]
```diff
@@ -18,44 +20,51 @@ private Categorical(Map<T, DoubleTensor> selectableValues) {
...
+        DoubleTensor sumOfProbabilities = getSumOfProbabilities(x.getShape());
+        if (!sumOfProbabilities.lessThanOrEqual(0.).allFalse()) {
             throw new IllegalArgumentException("Cannot sample from a zero probability setup.");
         }
-        final double probability = selectableValues.get(x).scalar() / sumOfProbabilities;
-        return DoubleTensor.scalar(Math.log(probability));
+
+        DoubleTensor logProb = DoubleTensor.zeros(x.getShape());
+        for (Map.Entry<T, DoubleTensor> entry : selectableValues.entrySet()) {
+            logProb.plusInPlace(x.equalsMask(entry.getKey()).timesInPlace(entry.getValue().div(sumOfProbabilities).logInPlace()));
```
Comment (migwellian) [inline keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Categorical.java]: In https://github.com/improbable-research/keanu/pull/265 I have added `elementwiseEquals(T value)`. I think you could use that here, and there'd be no need to create the `equalsMask` method.

File: keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Categorical.java [hunk 2]
```diff
@@ -3,10 +3,12 @@
 import java.util.Map;
 
 import io.improbable.keanu.distributions.Distribution;
+import io.improbable.keanu.tensor.bool.BooleanTensor;
 import io.improbable.keanu.tensor.dbl.DoubleTensor;
+import io.improbable.keanu.tensor.generic.GenericTensor;
 import io.improbable.keanu.vertices.dbl.KeanuRandom;
 
-public class Categorical<T> implements Distribution<T> {
+public class Categorical<T> implements Distribution<GenericTensor<T>> {
```
Comment (Pinpickle) [inline keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Categorical.java]: Is there a reason this has to be a `GenericTensor` as opposed to the interface `Tensor`?

File: keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Categorical.java [hunk 3]
```diff
@@ -18,44 +20,51 @@ private Categorical(Map<T, DoubleTensor> selectableValues) {
         this.selectableValues = selectableValues;
     }
 
-    public T sample(int[] shape, KeanuRandom random) {
-        double sumOfProbabilities = getSumOfProbabilities();
-        double p = random.nextDouble();
-        double sum = 0;
-
-        if (sumOfProbabilities == 0.0) {
+    public GenericTensor<T> sample(int[] shape, KeanuRandom random) {
```
Comment (Pinpickle) [inline keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Categorical.java]: At a high level I understand that this is the generalisation of the case of sampling a scalar (what it used to be). In the old version,  the logic was fairly penetrable in terms of understanding. 

In this new version, it's hard to see why this works. Is there anything we can do with breaking up into more functions, variable names and comments to make this more understandable? I'll add some more specific questions that come to mind when reading below.

File: keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Categorical.java [hunk 4]
```diff
@@ -18,44 +20,51 @@ private Categorical(Map<T, DoubleTensor> selectableValues) {
...
+            if (mask.allTrue()) {
                 break;
             }
         }
-        if (value == null) {
-            T[] values = (T[]) selectableValues.keySet().toArray();
-            value = values[values.length - 1];
+
+        if (value.isNull()) {
+            value = new GenericTensor<>((T[]) selectableValues.keySet().toArray(), shape);
```
Comment (Pinpickle) [inline keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Categorical.java]: Thoughts as a reader: Under what circumstances will this branch be taken? (i.e. when will value be null). Why is the value being set here sensible?

File: keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Categorical.java [hunk 5]
```diff
@@ -18,44 +20,51 @@ private Categorical(Map<T, DoubleTensor> selectableValues) {
...
+        GenericTensor<T> value = new GenericTensor<>(shape);
+
         for (Map.Entry<T, DoubleTensor> entry : selectableValues.entrySet()) {
-            sum += entry.getValue().scalar() / sumOfProbabilities;
-            if (p < sum) {
-                value = entry.getKey();
+            sum.plusInPlace(entry.getValue().div(sumOfProbabilities));
+
+            BooleanTensor mask = sum.greaterThan(p);
+            value.setWithMaskInPlace(mask.toDoubleMask(), entry.getKey());
```
Comment (Pinpickle) [inline keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Categorical.java]: Thoughts as a reader: Why does this work? Is there a consequence of setting this at every iteration, as opposed to once in the original version?

File: keanu-project/src/main/java/io/improbable/keanu/vertices/generic/probabilistic/discrete/CategoricalVertex.java [hunk 6]
```diff
@@ -54,36 +59,48 @@
         return copy;
     }
 
-    public CategoricalVertex(Map<T, DoubleVertex> selectableValues) {
+    public CategoricalVertex(int[] tensorShape, Map<T, DoubleVertex> selectableValues) {
+        checkTensorsMatchNonScalarShapeOrAreScalar(tensorShape, selectableValuesShapes(selectableValues));
```
Comment (Pinpickle) [inline keanu-project/src/main/java/io/improbable/keanu/vertices/generic/probabilistic/discrete/CategoricalVertex.java]: Should this check be done at the distribution level?

File: keanu-project/src/main/java/io/improbable/keanu/vertices/generic/probabilistic/discrete/CategoricalVertex.java [hunk 7]
```diff
@@ -17,7 +21,8 @@
 
 import static java.util.stream.Collectors.toMap;
 
-public class CategoricalVertex<T> extends Vertex<T> implements Probabilistic<T> {
+
+public class CategoricalVertex<T> extends Vertex<GenericTensor<T>> implements Probabilistic<GenericTensor<T>> {
```
Comment (Pinpickle) [inline keanu-project/src/main/java/io/improbable/keanu/vertices/generic/probabilistic/discrete/CategoricalVertex.java]: Do you think a javadoc on how this vertex works with tensors is useful? Or is it clear from context?

File: keanu-project/src/main/java/io/improbable/keanu/tensor/generic/GenericTensor.java [hunk 8]
```diff
@@ -73,6 +75,35 @@ public T getValue(int... index) {
         return this;
     }
 
+    public boolean isNull() {
+        return this.data == null;
+    }
+
+    public GenericTensor<T> setWithMaskInPlace(DoubleTensor mask, T value) {
```
Comment (Pinpickle) [inline keanu-project/src/main/java/io/improbable/keanu/tensor/generic/GenericTensor.java]: Awesome that we have a new util for this! Should this be generalised across all tensors?

File: keanu-project/src/main/java/io/improbable/keanu/tensor/generic/GenericTensor.java [hunk 9]
```diff
@@ -73,6 +75,35 @@ public T getValue(int... index) {
...
+        if (this.getLength() != mask.getLength()) {
+            throw new IllegalArgumentException("The lengths of the tensor and mask must match, but got tensor length: " + this.getLength() + ", mask length: " + mask.getLength());
+        }
+
+        if (data == null) {
+            data = (T[]) new Object[(int) this.getLength()];
```
Comment (Pinpickle) [inline keanu-project/src/main/java/io/improbable/keanu/tensor/generic/GenericTensor.java]: It's fine if we need this cast, but are we sure it won't cause problems? Casting arrays can fail in non-intuitive ways due to covariance: https://stackoverflow.com/questions/28570877/java-covariant-array-bad

File: keanu-project/src/test/java/io/improbable/keanu/tensor/generic/SimpleTensorTest.java [hunk 10]
```diff
@@ -23,7 +31,7 @@ public void canGetRandomAccessValue() {
                 Something.C, Something.D, Something.B,
                 Something.D, Something.A, Something.C
             },
-            new int[]{3, 3}
+            3, 3
```
Comment (Pinpickle) [inline keanu-project/src/test/java/io/improbable/keanu/tensor/generic/SimpleTensorTest.java]: 😍

File: keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Categorical.java [hunk 11]
```diff
@@ -18,44 +20,51 @@ private Categorical(Map<T, DoubleTensor> selectableValues) {
...
+        return logProb;
     }
 
-    private double getSumOfProbabilities() {
-        double sumP = 0.0;
+    private DoubleTensor getSumOfProbabilities(int[] shape) {
+        DoubleTensor sumP = DoubleTensor.zeros(shape);
         for (DoubleTensor p : selectableValues.values()) {
-            sumP += p.scalar();
+            sumP.plusInPlace(p);
```
Comment (Pinpickle) [inline keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Categorical.java]: Will the input shape always match the shape of the selectable values? If so, does it need to be an input? If not, should `sumP` be reassigned here as in some cases, when tensor shapes differ, a new tensor is created in `plusInPlace` (instead of reassigning in place)

File: keanu-project/src/test/java/io/improbable/keanu/vertices/generic/probabilistic/discrete/CategoricalVertexTest.java [hunk 12]
```diff
@@ -1,32 +1,140 @@
...
 
 import java.util.Arrays;
 import java.util.LinkedHashMap;
 import java.util.Map;
 import java.util.stream.Collectors;
 
 import static org.junit.Assert.assertEquals;
+import org.junit.rules.ExpectedException;
 
 public class CategoricalVertexTest {
```
Comment (Pinpickle) [inline keanu-project/src/test/java/io/improbable/keanu/vertices/generic/probabilistic/discrete/CategoricalVertexTest.java]: Awesome tests! Looks like you're catching all those permutations of possible shape variants which could have led to nasty edge cases.

File: keanu-project/src/test/java/io/improbable/keanu/tensor/generic/SimpleTensorTest.java [hunk 13]
```diff
@@ -106,4 +113,58 @@ public void canSliceRankTwoTensor() {
...
+            new Something[]{
+                Something.A, Something.B, Something.B,
+                Something.C, Something.D, Something.B,
+                Something.D, Something.A, Something.C
+            },
+            3, 3);
+        DoubleTensor mask = DoubleTensor.scalar(1);
+
+        thrown.expect(IllegalArgumentException.class);
+        thrown.expectMessage("The lengths of the tensor and mask must match, but got tensor length: " + tensor.getLength() + ", mask length: " + mask.getLength());
```
[the remark this answers is not in the document]
Comment (Pinpickle) [reply keanu-project/src/test/java/io/improbable/keanu/tensor/generic/SimpleTensorTest.java]: This method does make it clear exactly which line we are expecting to fail, though, and makes it impossible for other lines to pass the test. Trade-offs either way

File: keanu-project/src/test/java/io/improbable/keanu/vertices/generic/probabilistic/discrete/CategoricalVertexTest.java [hunk 14]
```diff
@@ -1,32 +1,140 @@
...
+
     @Before
     public void setup() {
         random = new KeanuRandom(1);
     }
 
+    @Test
+    public void cantCreateCategoricalVertexIfShapeIsNotSpecifiedAndNonScalarShapesDoNotMatch() {
+        DoubleTensor t1 = DoubleTensor.create(new double[] {0., 0.5, 0.8, 0.2}, 2, 2);
+        DoubleTensor t2 = DoubleTensor.create(new double[] {0., 0.5, 0.8, 0.2}, 4, 1);
```
Comment (migwellian) [inline keanu-project/src/test/java/io/improbable/keanu/vertices/generic/probabilistic/discrete/CategoricalVertexTest.java]: `DoubleTensor t2 = t1.duplicate()`

File: keanu-project/src/test/java/io/improbable/keanu/vertices/generic/probabilistic/discrete/CategoricalVertexTest.java [hunk 15]
```diff
@@ -1,32 +1,140 @@
...
+    public ExpectedException thrown = ExpectedException.none();
+
     @Before
     public void setup() {
         random = new KeanuRandom(1);
     }
 
+    @Test
+    public void cantCreateCategoricalVertexIfShapeIsNotSpecifiedAndNonScalarShapesDoNotMatch() {
+        DoubleTensor t1 = DoubleTensor.create(new double[] {0., 0.5, 0.8, 0.2}, 2, 2);
```
Comment (migwellian) [inline keanu-project/src/test/java/io/improbable/keanu/vertices/generic/probabilistic/discrete/CategoricalVertexTest.java]: You use this in many of the tests: you can make it a member variable.

File: keanu-project/src/test/java/io/improbable/keanu/vertices/generic/probabilistic/discrete/CategoricalVertexTest.java [hunk 16]
```diff
@@ -152,10 +261,10 @@ public void logProbOfCategoryIsEquivalentToItsLogProbabilityDividedBySum() {
 
         CategoricalVertex<TestEnum> select = new CategoricalVertex<>(selectableValues);
 
-        assertEquals(Math.log(probA / total), select.logProb(TestEnum.A), 1e-6);
-        assertEquals(Math.log(probB / total), select.logProb(TestEnum.B), 1e-6);
-        assertEquals(Math.log(probC / total), select.logProb(TestEnum.C), 1e-6);
-        assertEquals(Math.log(probD / total), select.logProb(TestEnum.D), 1e-6);
+        assertEquals(Math.log(probA / total), select.logProb(new GenericTensor<>(TestEnum.A)), 1e-6);
```
Comment (migwellian) [inline keanu-project/src/test/java/io/improbable/keanu/vertices/generic/probabilistic/discrete/CategoricalVertexTest.java]: you could add a helper method `CategoricalVertex#logProb(T)`

File: keanu-project/src/test/java/io/improbable/keanu/tensor/generic/SimpleTensorTest.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (migwellian) [reply keanu-project/src/test/java/io/improbable/keanu/tensor/generic/SimpleTensorTest.java]: That's a good point

File: keanu-project/src/main/java/io/improbable/keanu/tensor/Tensor.java [hunk 17]
```diff
@@ -40,6 +43,22 @@ static BooleanTensor elementwiseEquals(Tensor a, Tensor b) {
...
+    static IntegerTensor create(Integer value, int[] shape) {
+        return IntegerTensor.create(value, shape);
+    }
+
+    static BooleanTensor create(Boolean value, int[] shape) {
+        return BooleanTensor.create(value, shape);
+    }
+
+    static <T> GenericTensor<T> create(T value, int[] shape) {
+        return new GenericTensor<>(value, shape);
```
Comment (migwellian) [inline keanu-project/src/main/java/io/improbable/keanu/tensor/Tensor.java]: for consistency with the other Tensor types, we should probably also have a `GenericTensor.create` method

File: keanu-project/src/test/java/io/improbable/keanu/vertices/generic/probabilistic/discrete/CategoricalVertexTest.java (hunk 14 above)
Comment (icantsayrural) [reply keanu-project/src/test/java/io/improbable/keanu/vertices/generic/probabilistic/discrete/CategoricalVertexTest.java]: Not an exact duplicate, but I can reshape it.

File: keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Categorical.java [hunk 18]
```diff
@@ -18,44 +20,51 @@ private Categorical(Map<T, DoubleTensor> selectableValues) {
         this.selectableValues = selectableValues;
```
[the remark this answers is not in the document]
Comment (icantsayrural) [reply keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Categorical.java]: As discussed offline - right now we keep tensor shape validations in vertices so I will do the same in this PR.

File: keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Categorical.java [hunk 19]
```diff
@@ -18,44 +20,61 @@ private Categorical(Map<T, DoubleTensor> selectableValues) {
...
+            if (sampleValuesSetSoFar.allTrue()) {
                 break;
             }
         }
-        if (value == null) {
+
+        if (!sampleValuesSetSoFar.allTrue()) {
             T[] values = (T[]) selectableValues.keySet().toArray();
-            value = values[values.length - 1];
+            sample = Tensor.create(values[values.length - 1], shape);
```
Comment (migwellian) [inline keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Categorical.java]: You can do it without the cast if you avoid raw arrays:

            Set<T> values = selectableValues.keySet();
            T lastValue = ImmutableList.copyOf(values).get(values.size() - 1);
            sample = Tensor.create(lastValue, shape);

File: keanu-project/src/main/java/io/improbable/keanu/tensor/Tensor.java [hunk 20]
```diff
@@ -31,13 +31,17 @@ static BooleanTensor elementwiseEquals(Tensor a, Tensor b) {
...
     static <T> Tensor<T> scalar(T value) {
-        return new GenericTensor<>(value);
+        return GenericTensor.scalar(value);
     }
 
     static <T> Tensor<T> placeHolder(long[] shape) {
         return new GenericTensor<>(shape);
     }
 
+    static <T> GenericTensor<T> create(T value, long[] shape) {
```
Comment (migwellian) [inline keanu-project/src/main/java/io/improbable/keanu/tensor/Tensor.java]: We should also add specific versions , i.e.

    static DoubleTensor create(Double value, long[] shape) {
        return DoubleTensor.create(value, shape);
    }

    static BooleanTensor create(Boolean value, long[] shape) {
        return BooleanTensor.create(value, shape);
    }

    static IntegerTensor create(Integer value, long[] shape) {
        return IntegerTensor.create(value, shape);
    }

File: keanu-project/src/main/java/io/improbable/keanu/tensor/bool/SimpleBooleanTensor.java [hunk 21]
```diff
@@ -147,13 +152,13 @@ public BooleanTensor setBooleanIf(BooleanTensor trueValue, BooleanTensor falseVa
...
     @Override
-    public <T> Tensor<T> setIf(Tensor<T> trueValue, Tensor<T> falseValue) {
+    public <T> Tensor<T> where(Tensor<T> trueValue, Tensor<T> falseValue) {
         FlattenedView<T> trueValuesFlattened = trueValue.getFlattenedView();
         FlattenedView<T> falseValuesFlattened = falseValue.getFlattenedView();
 
         T[] result = (T[]) (new Object[data.length]);
         for (int i = 0; i < result.length; i++) {
-            result[i] = data[i] ? trueValuesFlattened.get(i) : falseValuesFlattened.get(i);
+            result[i] = data[i] ? trueValuesFlattened.getOrScalar(i) : falseValuesFlattened.getOrScalar(i);
```
Comment (migwellian) [inline keanu-project/src/main/java/io/improbable/keanu/tensor/bool/SimpleBooleanTensor.java]: Nice. Was this a bug? We should add an explicit test for the scalar case.

File: keanu-project/src/main/java/io/improbable/keanu/tensor/generic/GenericTensor.java [hunk 22]
```diff
@@ -22,6 +22,18 @@
     private long[] shape;
     private long[] stride;
 
+    public static <T> GenericTensor<T> create(T data, long[] shape) {
+        if (Arrays.equals(shape, Tensor.SCALAR_SHAPE)) {
```
Comment (migwellian) [inline keanu-project/src/main/java/io/improbable/keanu/tensor/generic/GenericTensor.java]: I don't think you need this branch - you can always `return new GenericTensor<>(data, shape);`

File: keanu-project/src/main/java/io/improbable/keanu/tensor/generic/GenericTensor.java [hunk 23]
```diff
@@ -32,6 +44,14 @@ public GenericTensor(T[] data, long[] shape) {
         }
     }
 
+    public GenericTensor(T data, long[] shape) {
```
Comment (migwellian) [inline keanu-project/src/main/java/io/improbable/keanu/tensor/generic/GenericTensor.java]: we already have `public GenericTensor(long[] shape, T value)`

File: keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Categorical.java (hunk 19 above)
Comment (icantsayrural) [reply keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Categorical.java]: Cool!

File: keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Categorical.java [hunk 24]
```diff
@@ -18,44 +20,51 @@ private Categorical(Map<T, DoubleTensor> selectableValues) {
...
-        T value = null;
+        double p = random.nextDouble();
+        DoubleTensor sum = DoubleTensor.zeros(shape);
+        GenericTensor<T> value = new GenericTensor<>(shape);
+
         for (Map.Entry<T, DoubleTensor> entry : selectableValues.entrySet()) {
-            sum += entry.getValue().scalar() / sumOfProbabilities;
-            if (p < sum) {
-                value = entry.getKey();
+            sum.plusInPlace(entry.getValue().div(sumOfProbabilities));
```
[the remark this answers is not in the document]
Comment (icantsayrural) [reply keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Categorical.java]: What do you mean?
